core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java

```diff
@@ -0,0 +1,367 @@
+     * {@link NoClassDefFoundError} errors happens, rather than the class itself.
+     * @param events events collected in this telemetry window.
+     * @return the events formatted in a map.
+     */
+    @Nonnull
+    private JSONArray formatEvents(@Nonnull ConcurrentHashMap<List<StackTraceElement>, MissingClassEvent> events) {
+        JSONArray jsonEvents = new JSONArray();
+
+        events.forEach((stackTrace, event) -> {
+            JSONObject eventObject = new JSONObject();
+            eventObject.put("className", event.getClassName());
+            eventObject.put("class", event.getClassName());
+            eventObject.put("time", event.getTime());
+            eventObject.put("occurrences", Long.toString(event.getOccurrences()));
+            eventObject.put("stacktrace", event.getStackTrace());
+
+            jsonEvents.add(eventObject);
+        });
+
+        return jsonEvents;
+    }
+
+    /**
+     * The current time in the same way as other telemetry implementations.
+     * @return the UTC time formatted with the pattern yyyy-MM-dd'T'HH:mm'Z'
+     */
+    @Nonnull
+    static String clientDateString() {
+        TimeZone tz = TimeZone.getTimeZone("UTC");
+        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'");
+        df.setTimeZone(tz); // strip timezone
+        return df.format(new Date());
+    }
+
+    /**
+     * Store the exception if it's from a split package of Java. This method report this exception directly, it doesn't
+     * look into the causes or suppressed exceptions of the exception specified. This method tends to be used in the
+     * ClassLoader directly. Outside the class loaders is best to use {@link #reportExceptionInside(Throwable)}
+     * @param name the name of the class
+     * @param e the throwable to report if needed
+     */
+    public static void reportException(@Nonnull String name, @Nonnull Throwable e) {
+        if (enabled()) {
+            //ClassDefFoundError uses / instead of .
+            name = name.replace('/', '.').trim();
+
+            // We call the methods in this order because if the missing class is not java related, we don't loop over the
+            // stack trace to look if it's not thrown from an ignored place avoiding an impact on performance.
+            if (isFromMovedPackage(name) && !calledFromIgnoredPlace(e)) {
+                events.put(name, e);
+                if (LOGGER.isLoggable(Level.WARNING))
+                    LOGGER.log(Level.WARNING, "Added a missed class for missing class telemetry. Class: " + name, e);
+            }
+        }
+    }
+
+    /**
+     * Determine if the exception specified was thrown from an ignored place
+     * @param throwable The exception thrown
+     * @return true if in the stack trace there is an ignored method / class.
+     */
+    private static boolean calledFromIgnoredPlace(@Nonnull Throwable throwable) {
+        for(String[] ignoredPlace : IGNORED_PLACES) {
```

Comment:
Java 8-like method using lambda?

Reply:
sorry, where is there a lambda?

Reply:
It's not, that's why I was suggesting maybe using it.

Reply:
Because it's executed in every CNFE to check whether the exception should be sent, I opted for this approach to avoid impacting the performance.